interfaces/system_key.go

```diff
@@ -20,47 +20,86 @@
 package interfaces
 
 import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
 	"os"
 	"path/filepath"
-
-	"gopkg.in/yaml.v2"
+	"reflect"
+	"strings"
 
 	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/logger"
 	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/release"
 )
 
+// ErrSystemKeyIncomparableVersions indicates that the system-key
+// on disk and the system-key calculated from generateSystemKey
+// have different inputs and are therefor incomparable.
+//
+// This means:
+// - "snapd" needs to re-generate security profiles
+// - "snap run" cannot wait for those security profiles
+var (
+	ErrSystemKeyIncomparableVersions = errors.New("system-key versions not comparable")
+	ErrSystemKeyMissing              = errors.New("system-key missing on disk")
+)
+
 // systemKey describes the environment for which security profiles
 // have been generated. It is useful to compare if the current
 // running system is similar enough to the generated profiles or
 // if the profiles need to be re-generated to match the new system.
+//
+// Note that this key gets generated on *each* `snap run` - so it
+// *must* be cheap to calculate it (no hashes of big binaries etc).
 type systemKey struct {
-	BuildID          string   `yaml:"build-id"`
-	AppArmorFeatures []string `yaml:"apparmor-features"`
-	NFSHome          bool     `yaml:"nfs-home"`
-	OverlayRoot      string   `yaml:"overlay-root"`
-	Core             string   `yaml:"core,omitempty"`
-	SecCompActions   []string `yaml:"seccomp-features"`
+	// IMPORTANT: when adding new inputs bump this version
+	Version int `json:"version"`
+
+	BuildID          string   `json:"build-id"`
+	AppArmorFeatures []string `json:"apparmor-features"`
+	NFSHome          bool     `json:"nfs-home"`
+	OverlayRoot      string   `json:"overlay-root"`
+	SecCompActions   []string `json:"seccomp-features"`
 }
 
 var (
 	isHomeUsingNFS  = osutil.IsHomeUsingNFS
 	mockedSystemKey *systemKey
 )
 
+func findSnapdPath() (string, error) {
```

Comment:
not for now but I think we have various functions like this, maybe we can find a place for them? dirs?

Reply:
Good point. I will see if I can squeeze it in here or into a followup.